File: ADT-20240508/ad-t/src/bin/h.rs
use proconio::input;
use std::collections::VecDeque;

fn main() {
    input! {
        n:usize,m:usize,
        ab:[(usize,usize);m],
        q:usize,
        xk:[(usize,usize);q],
    }

    let mut edges: Vec<Vec<usize>> = vec![vec![]; n];

    for (a, b) in ab {
        edges[a - 1].push(b - 1);
        edges[b - 1].push(a - 1);
    }

    let mut dis: Vec<isize> = vec![-1; n];

    for (mut x, k) in xk {
        x -= 1;
        let mut que = VecDeque::new();
        dis[x] = 0;
        let mut vs: Vec<usize> = vec![];

        while let Some(u) = que.pop_front() {
            vs.push(u);
            if dis[u] == k as isize {
                continue;
            }

            for j in 0..edges[u].len() {
                let v = edges[u][j];
                if dis[v] != -1 {
                    continue;
                }
                dis[v] = dis[u] + 1;
                que.push_back(v);
            }
        }
        let mut ans = 0;
        for j in 0..vs.len() {
            ans += vs[j] + 1;
            dis[vs[j]] = -1;
        }
        println!("{}", ans);
    }
}

File: ADT-20240508/ad-t/h.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    
	int N,M;
	cin>>N>>M;
	vector<vector<int>> E(N);
	for(int i=0;i<M;i++){
		int a,b;
		cin>>a>>b;
		a--,b--;
		E[a].push_back(b);
		E[b].push_back(a);
	}
	
	vector<int> dis(N,-1);
	
	int Q;
	cin>>Q;
	
	for(int i=0;i<Q;i++){
		int x,k;
		cin>>x>>k;
		x--;
		queue<int> q;
		q.push(x);
		dis[x] = 0;
		vector<int> vs;
		while(q.size()>0){
			int u = q.front();
			q.pop();
			vs.push_back(u);
			if(dis[u]==k){
				continue;
			}
			for(int j=0;j<E[u].size();j++){
				int v = E[u][j];
				if(dis[v]!=-1)continue;
				dis[v] = dis[u] + 1;
				q.push(v);
			}
		}
		int ans = 0;
		for(int j=0;j<vs.size();j++){
			ans += vs[j]+1;
			dis[vs[j]] = -1;
		}
		cout<<ans<<endl;
	}
	
    return 0;
}
